File: cpp/src/token.cpp
#include "token.hpp"

namespace wildling {

int Token::default_integer(std::optional<int> option, int fallback) {
    return option.has_value() && option.value() >= 0 ? option.value() : fallback;
}

int Token::pow_int(int base, int exp) {
    int result = 1;
    for (int i = 0; i < exp; ++i) {
        result *= base;
    }
    return result;
}

Token::Token(TokenOptions options)
    : src_(std::move(options.src)),
      start_length_(default_integer(options.start_length, 1)),
      end_length_(default_integer(options.end_length, 1)),
      variants_(std::move(options.variants)),
      count_(0) {
    for (int length = start_length_; length <= end_length_; ++length) {
        count_ += pow_int(static_cast<int>(variants_.size()), length);
    }
}

int Token::count() const {
    return count_;
}

const std::string& Token::src() const {
    return src_;
}

std::string Token::get(int index) const {
    if (index > count_ - 1 || index < 0) {
        return "";
    }
    if (index == 0 && start_length_ == 0) {
        return "";
    }

    int index_with_offset = index;
    int string_length = start_length_;
    for (string_length = start_length_; string_length <= end_length_; ++string_length) {
        int offset_count = pow_int(static_cast<int>(variants_.size()), string_length);
        if (index_with_offset < offset_count) {
            break;
        }
        index_with_offset -= offset_count;
    }

    std::string out;
    out.reserve(static_cast<size_t>(string_length) * 4);
    for (int i = 0; i < string_length; ++i) {
        int variant_index = index_with_offset % static_cast<int>(variants_.size());
        index_with_offset /= static_cast<int>(variants_.size());
        out += variants_[static_cast<size_t>(variant_index)];
    }
    return out;
}

}  // namespace wildling

File: rust/src/token.rs
#[derive(Clone, Debug)]
pub struct TokenOptions {
    pub string: Option<String>,
    pub start_length: Option<i32>,
    pub end_length: Option<i32>,
    pub variants: Vec<String>,
    pub src: String,
}

#[derive(Clone, Debug)]
pub struct Token {
    src: String,
    start_length: i32,
    end_length: i32,
    variants: Vec<String>,
    count: i32,
}

fn default_integer(option: Option<i32>, fallback: i32) -> i32 {
    match option {
        Some(v) if v >= 0 => v,
        _ => fallback,
    }
}

fn pow_int(base: i32, exp: i32) -> i32 {
    let mut result = 1;
    for _ in 0..exp {
        result *= base;
    }
    result
}

impl Token {
    pub fn new(options: TokenOptions) -> Self {
        let start_length = default_integer(options.start_length, 1);
        let end_length = default_integer(options.end_length, 1);
        let variants = options.variants;
        let mut count = 0;
        let mut length = start_length;
        while length <= end_length {
            count += pow_int(variants.len() as i32, length);
            length += 1;
        }
        Self {
            src: options.src,
            start_length,
            end_length,
            variants,
            count,
        }
    }

    pub fn count(&self) -> i32 {
        self.count
    }

    pub fn src(&self) -> &str {
        &self.src
    }

    pub fn get(&self, index: i32) -> String {
        if index > self.count - 1 || index < 0 {
            return String::new();
        }
        if index == 0 && self.start_length == 0 {
            return String::new();
        }

        let mut index_with_offset = index;
        let mut string_length = self.start_length;
        let mut length = self.start_length;
        while length <= self.end_length {
            let offset_count = pow_int(self.variants.len() as i32, length);
            if index_with_offset < offset_count {
                string_length = length;
                break;
            }
            index_with_offset -= offset_count;
            length += 1;
        }

        let mut out = String::new();
        for _ in 0..string_length {
            let variant_index = (index_with_offset as usize) % self.variants.len();
            index_with_offset /= self.variants.len() as i32;
            out.push_str(&self.variants[variant_index]);
        }
        out
    }
}
